pkg/controllers/nodepool/hash/controller.go

```diff
@@ -37,19 +38,35 @@ var _ operatorcontroller.TypedController[*v1beta1.NodePool] = (*Controller)(nil)
 // Controller is hash controller that constructs a hash based on the fields that are considered for static drift.
 // The hash is placed in the metadata for increased observability and should be found on each object.
 type Controller struct {
-	kubeClient client.Client
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
 }
 
-func NewController(kubeClient client.Client) operatorcontroller.Controller {
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider) operatorcontroller.Controller {
 	return operatorcontroller.Typed[*v1beta1.NodePool](kubeClient, &Controller{
-		kubeClient: kubeClient,
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
 	})
 }
 
 // Reconcile the resource
 func (c *Controller) Reconcile(ctx context.Context, np *v1beta1.NodePool) (reconcile.Result, error) {
 	stored := np.DeepCopy()
 
+	// To avoid a breaking change on the NodePool API, we will be setting a default APIVersion and Kind
+	// defined by the cloudprovider to each nodeClassRef on every NodePool. This will be removed once
+	// the NodePool API requires APIVersion and Kind to be set at NodePool creation.
+	// TODO: remove at v1 when APIVersion and Kind are required fields on NodePool
```

Comment:
I was thinking here, should we make it required, or should we make it optional and just require that cloud providers add the defaulting logic to include the GVK into NodePools? If we require it, we can't default it. Is there a point in allowing users to be wrong here?